pkg/yurttunnel/server/serveraddr/addr.go

```diff
@@ -0,0 +1,283 @@
+/*
+Copyright 2020 The OpenYurt Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package serveraddr
+
+import (
+	"errors"
+	"fmt"
+	"net"
+
+	"github.com/alibaba/openyurt/pkg/projectinfo"
+	"github.com/alibaba/openyurt/pkg/yurttunnel/constants"
+
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+)
+
+// GetServerAddr gets the service address that exposes the tunnel server for
+// tunnel agent to connect
+func GetTunnelServerAddr(clientset kubernetes.Interface) (string, error) {
+	var (
+		ip      net.IP
+		host    string
+		tcpPort int32
+	)
+
+	// get tunnel server resources
+	svc, eps, nodeLst, err := getTunnelServerResources(clientset)
+	if err != nil {
+		return "", err
+	}
+
+	dnsNames, ips, err := extractTunnelServerDNSandIPs(svc, eps, nodeLst)
+	if err != nil {
+		return "", err
+	}
+
+	for _, tmpIP := range ips {
+		// we use the first non-loopback IP address.
+		if tmpIP.String() != "127.0.0.1" {
+			ip = tmpIP
+			break
+		}
+	}
+
+	if ip == nil {
+		if len(dnsNames) == 0 {
```

Comment:
may be we can remove unnecessary else clause?
```go
if len(dnsNames) == 0 {
    return "", errors.New("there is no available ip")
}
host = dnsNames[0]
```

Reply:
fixed